File: leetcode/0205-isomorphic-strings/solution.cpp
class Solution {
public:
    /**
     * RUNTIME: 6 ms (91.3%)
     * MEMORY: 7 MB (75.6%)
    */
    bool isIsomorphic(string s, string t) {
        if (s.size() != t.size()) { return false; }
        unordered_map<int, int> hashmap;
        vector<int> values;
        for (int i=0; i < s.size(); i++) {
            if (hashmap.find(s[i]) == hashmap.end()) {
                if ( std::find(values.begin(), values.end(), t[i])  != values.end() ) {
                    return false;
                }
                hashmap[s[i]] = t[i];
                values.push_back(t[i]);
            }
            else {
                if (hashmap[s[i]] != t[i]) {
                    return false;
                }
            }
        }
        return true;
    }
};

File: leetcode/0205-isomorphic-strings/solution.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        """
        RUNTIME: 54 ms (80.12%)
        MEMORY: 14.2 MB (89.54%)
        """
        if len(s) != len(t): return False
        hashmap = {}
        for i in range(len(s)):
            if s[i] not in hashmap:
                if t[i] in hashmap.values():
                    return False
                hashmap[s[i]] = t[i]
            else:
                if hashmap[s[i]] != t[i]:
                    return False
        return True
